### PR TITLE
Create batches for NormalizeSubmissionPoints

### PR DESCRIPTION
**Pull request by MTsatsarov**
Closes https://github.com/SoftUni-Internal/exam-systems-issues/issues/1026

Added batches and separate TransactionScopes for BE job to normalize submissions and participants points.

### CODE REVIEW

File: Open Judge System/Services/OJS.Services/Business/ParticipantScores/ParticipantScoresBusinessService.cs [hunk 1]
```diff
@@ -54,13 +54,8 @@ public void RecalculateForParticipantByProblem(int participantId, int problemId)
 
         public void NormalizeAllPointsThatExceedAllowedLimit()
         {
-            using (var scope = TransactionsHelper.CreateLongRunningTransactionScope())
-            {
                 this.NormalizeSubmissionPoints();
                 this.NormalizeParticipantScorePoints();
```
Comment (gogo4ds) [inline Open Judge System/Services/OJS.Services/Business/ParticipantScores/ParticipantScoresBusinessService.cs:58]: These operations should be in one transaction

File: Open Judge System/Services/OJS.Services/Business/ParticipantScores/ParticipantScoresBusinessService.cs [hunk 2]
```diff
@@ -377,37 +372,60 @@ private double CalculateTimeInMinutesFromParticipationStartToLastSubmission(Part
...
+                    {
+                        x.Submission.Points = x.ProblemMaxPoints;
 
-                    this.submissionsData.Update(x.Submission);
-                });
+                        this.submissionsData.Update(x.Submission, true);
+                    });
+                    
+                    this.submissionsData.SaveChanges();
+                    scope.Complete();
```
Comment (gogo4ds) [inline Open Judge System/Services/OJS.Services/Business/ParticipantScores/ParticipantScoresBusinessService.cs:399]: There is no point of executing this in a transaction scope as `this.submissionsData.SaveChanges();` will be a transaction for all the updates before it.

File: Open Judge System/Services/OJS.Services/Business/ParticipantScores/ParticipantScoresBusinessService.cs [hunk 3]
```diff
@@ -377,37 +372,60 @@ private double CalculateTimeInMinutesFromParticipationStartToLastSubmission(Part
...
-                        x.ProblemMaxPoints,
-                        x.Particinapnt));
+                itemsToSkip += itemsToTake;
+            }
+        }
+
+
+        private void NormalizeParticipantScorePoints()
+        {
+            using (var scope = TransactionsHelper.CreateLongRunningTransactionScope())
```
Comment (gogo4ds) [inline Open Judge System/Services/OJS.Services/Business/ParticipantScores/ParticipantScoresBusinessService.cs:409]: Here is the same..

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Open Judge System/Services/OJS.Services/Business/ParticipantScores/ParticipantScoresBusinessService.cs (hunk 1 above)
Comment (MTsatsarov) [reply Open Judge System/Services/OJS.Services/Business/ParticipantScores/ParticipantScoresBusinessService.cs:58]: After the last discussion, it was decided to remove the transactions. and submissions and Participant scores updates to be done separately

File: Open Judge System/Services/OJS.Services/Business/ParticipantScores/ParticipantScoresBusinessService.cs (hunk 2 above)
Comment (MTsatsarov) [reply Open Judge System/Services/OJS.Services/Business/ParticipantScores/ParticipantScoresBusinessService.cs]: fixed

File: Open Judge System/Services/OJS.Services/Business/ParticipantScores/ParticipantScoresBusinessService.cs (hunk 3 above)
Comment (MTsatsarov) [reply Open Judge System/Services/OJS.Services/Business/ParticipantScores/ParticipantScoresBusinessService.cs]: fixed

File: Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs [hunk 4]
```diff
@@ -167,6 +167,23 @@ public void Delete(IEnumerable<ParticipantScore> participantScoresEnumerable)
...
+            //The submission TotalScoreSnapshotModifiedOn must be changed only if it is new submission in other way the results will not be ordered correctly.
+            bool shouldUpdateTotalScoreDate = submissionId != null && submissionId != participantScore.SubmissionId;
+            UpdateTotalScoreSnapshot(
+                participant, 
+                participantScore.Points, 
+                submissionPoints,
+                shouldUpdateTotalScoreDate);
+
+            participantScore.SubmissionId = submissionId;
+            participantScore.Points = submissionPoints;
```
Comment (gogo4ds) [inline Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs:182]: This is repetition of knowledge (from the overload above), which we can easily avoid.

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs (hunk 4 above)
Comment (MTsatsarov) [reply Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs]: Fixed

**Review by gogo4ds — DISMISSED** [review]

**Review by gogo4ds — APPROVED** [review]

**Pull request merged**